src/libraries/System.Formats.Tar/tests/TarEntry/GnuTarEntry.Tests.cs

```diff
@@ -93,5 +94,343 @@ public void SupportedEntryType_Fifo()
+            // * 302 bytes of padding
+            // * 512 bytes of the regular long path tar header
+            // * 1234 bytes for the data section containing the full long path
+            // * 302 bytes of padding
+            // Then it writes the actual regular file entry, containing:
+            // * 512 bytes of the regular tar header
+            // Totalling 4608.
+            // The regular file data section starts on the next byte.
+            long expectedDataOffset = canSeek ? 4609 : -1;
+            Assert.Equal(expectedDataOffset, actualEntry.DataOffset);
+        }
+        
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task DataOffset_RegularFile_LongLin_LongPath_Async(bool canSeek)
+        {
+            await using MemoryStream ms = new();
+            await using (TarWriter writer = new(ms, leaveOpen: true))
+            {
+                string veryLongName = new string('a', 1234); // Forces using a GNU LongPath entry
+                GnuTarEntry entry = new GnuTarEntry(TarEntryType.SymbolicLink, veryLongName);
+                entry.LinkName = new string('a', 1234); // Forces using a GNU LongLink entry
+                await writer.WriteEntryAsync(entry);
+            }
+            ms.Position = 0;
+
+            await using Stream streamToRead = new WrappedStream(ms, canWrite: true, canRead: true, canSeek: canSeek);
+            await using TarReader reader = new(streamToRead);
+            TarEntry actualEntry = await reader.GetNextEntryAsync();
+            Assert.NotNull(actualEntry);
+
+            // GNU first writes the long link and long path entries, containing:
+            // * 512 bytes of the regular long link tar header
+            // * 1234 bytes for the data section containing the full long link
+            // * 302 bytes of padding
+            // * 512 bytes of the regular long path tar header
+            // * 1234 bytes for the data section containing the full long path
+            // * 302 bytes of padding
+            // Then it writes the actual regular file entry, containing:
+            // * 512 bytes of the regular tar header
+            // Totalling 4608.
+            // The regular file data section starts on the next byte.
+            long expectedDataOffset = canSeek ? 4609 : -1;
+            Assert.Equal(expectedDataOffset, actualEntry.DataOffset);
+        }
+
+        [Fact]
+        public void DataOffset_BeforeAndAfterArchive()
+        {
+            GnuTarEntry entry = new GnuTarEntry(TarEntryType.RegularFile, InitialEntryName);
+            Assert.Equal(-1, entry.DataOffset);
+            entry.DataStream = new MemoryStream();
+            entry.DataStream.WriteByte(5);
+
+            using MemoryStream ms = new();
+            using TarWriter writer = new(ms);
+            writer.WriteEntry(entry);
+            Assert.Equal(513, entry.DataOffset);
```

Comment:
What if we re-use the entry to write it again?  Can we make sure that the offset is updated correctly?  (same for async)

Reply:
I added tests that confirm that when we create an entry from scratch, the offset returns -1 as there is no archive stream yet. Then when the entry is written to a tar archive, the entry gets its offset updated to the correct location relative to the archive stream where it was written. This means that _writing_ an entry does update the offset.

I think for the case you're describing (getting an entry from an archive, then writing it into another archive) might not work, and the user would have to clone the entry into another one (hence why we have constructors to copy one entry to another). I can test this to double check, but I can tell you that writing an entry to an archive does update its offset already.

Reply:
I was just imagining someone writing the same entry to the archive multiple times.  I think that's possible since the type is mutable?

Reply:
Added a test to confirm that when writing an entry two times to the same archive will update the offset to the latest written position.